compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -195,7 +195,9 @@ public BType checkType(BLangExpression expr,
                            BType actualType,
                            BType expType,
                            DiagnosticCode diagCode) {
-        expr.setBType(checkType(expr.pos, actualType, expType, diagCode));
+        expr.setDeterminedType(actualType);
+        expr.setTypeCheckedType(checkType(expr.pos, actualType, expType, diagCode));
```

Comment:
Are there places that we set types that is not `symtable.semanticError` when something is not right with the type-checking of that statement/expression.
If it's always `semanticError` we can merge this 2 methods.

Reply:
Yes. Supporting that scenario is the reason for this separation. For example, consider a function `foo()` which returns a string. And now consider the following:
```ballerina
int x = foo();
```
Here, since the LHS and RHS are incompatible, we set `semanticError` as the type of the function call expr `foo()`. But in reality, we can (and we do) determine a type correctly for `foo()`. Being able to get this type through the semantic API was the original requirement. But since we overwrite that type to `semanticError` due to the type incompatibility, previously this info was lost.

Reply:
Oh I was thinking about checking for semanticError, but I don't think we have access to that here.
`this.determinedType != symtable.semanticError`

Reply:
So I'm fine with this change

Reply:
In this particular instance, yes, `this.determinedType != symtable.semanticError`. But there can be instances where the determined type is `semanticError`. For example, I think the following binary expr would be one such instance: `10 + 2.3`